### DISCUSSION TITLE
Create FK to email column in auth.users

### DISCUSSION THREAD

**Discussion by user72356** [supabase/supabase#11680] [Questions] [answered]
I'm trying to create a profile table that adds some personal user info to the auth.users table and trying to enforce integrity of the email column. There is a partial unique index on the auth.users.email column and despite my efforts I can't seem to create a FK.

I'm getting this error: "there is no unique constraint matching given keys for referenced table "users""

See below:

```
CREATE TABLE public.profile
(
    user_id uuid NOT NULL,
    email text NOT NULL,
    "firstName" text,
    "lastName" text,
    is_sso_user boolean NOT NULL generated always as (false) stored
);

ALTER TABLE ONLY public.profile
    ADD CONSTRAINT profile_pkey PRIMARY KEY (user_id);

ALTER TABLE ONLY public.profile
    ADD CONSTRAINT profile_unique_email UNIQUE (email, is_sso_user);

ALTER TABLE ONLY public.profile
    ADD CONSTRAINT profile_fkey_user_id FOREIGN KEY (user_id) REFERENCES auth.users(id);

-- why is this not working?
ALTER TABLE ONLY public.profile
    ADD CONSTRAINT profile_fkey_email FOREIGN KEY (email, is_sso_user) REFERENCES auth.users(email, is_sso_user);
```

**Comment by GaryAustin1** [comment, marked as answer]
The foreign key with two columns requires both columns be in a unique constraint on the table it is references.

**Reply by user72356** [reply]
Since I can't really modify auth.users, what would be the right approach to creating a foreign key on email? Since it has a partial unique index. I guess I might not be able to!

**Reply by GaryAustin1** [reply]
I'm not an expert, but can't think of a way to use foreign key.  

Why do you need the foreign key if you have a foreign key for user id?
I assume that email gets inserted from an auth.users trigger function in your profile table when the user is created.  You can also have a trigger on update of auth.users to keep it updated.

If you want to protect that column from user updates directly either put your protected columns in a separate secure profile table only updated by trigger (ie RLS blocks updates) OR have an update trigger on profiles that sets new.email = old.email so it can't be changed.  (Probably would need to have a check for the update coming from the auth.trigger (like no authenticated role) to make sure it can still update email if doing the 2nd approach.

**Reply by user72356** [reply]
Long story short, due to tech considerations outside of Supabase, my DB is sharded in such a way that the profile table is siloed out of the rest of the DB. So the profile table can't be joined up with the auth.users table in a query to retrieve email. email needs to be duplicated into profile. 

All of this happens in the app layer and I was looking for a way to guarantee the integrity of data at the DB level in order to be extra safe. I'm aware of triggers but it's still one level above DB constraints (and personally I prefer not to rely on triggers anyway for portability reasons). Not a problem in 99% of cases since I can create FKs on that duplicated data, but this is the first time I'm encountering a partial unique index.

I guess we're just going to have to let this one go through the cracks, I don't see another way short of modifying the auth.users table. I think this ability was removed recently.

**Reply by user72356** [reply]
BTW I just want to take a moment to thank you for answering so many of my questions over the past few months. Especially as you appear not to be employed by Supabase. Cheers man!
